sys/net/gnrc/link_layer/lwmac/lwmac.c

```diff
@@ -0,0 +1,745 @@
+        break;
+
+    /* Receiving incoming data */
+    case RECEIVING:
+        rtt_handler(LWMAC_EVENT_RTT_PAUSE); /* No duty cycling while TXing */
+        _set_netdev_state(&lwmac, NETOPT_STATE_IDLE);  /* Power up netdev */
+        break;
+
+    case STOPPED:
+        _set_netdev_state(&lwmac, NETOPT_STATE_OFF);
+        break;
+
+    /*********************** Control states ***********************************/
+    case START:
+        rtt_handler(LWMAC_EVENT_RTT_START);
+        lwmac_set_state(LISTENING);
+        break;
+
+    case STOP:
+        rtt_handler(LWMAC_EVENT_RTT_STOP);
+        lwmac_set_state(STOPPED);
+        break;
+
+    case RESET:
+        LOG_WARNING("Reset not yet implemented\n");
+        lwmac_set_state(STOP);
+        lwmac_set_state(START);
+        break;
+
+    /**************************************************************************/
+    default:
+        LOG_DEBUG("No actions for entering state %u\n", newstate);
+        return;
+    }
+
+    lwmac_schedule_update();
+}
+
+/******************************************************************************/
+
+/* Main state machine. Call whenever something happens */
+// TODO: Don't use global variables
+bool lwmac_update(void)
+{
+    lwmac.needs_rescheduling = false;
+
+    switch(lwmac.state)
+    {
+    case SLEEPING:
+
+        /* If a packet is scheduled, no other (possible earlier) packet can be
+         * sent before the first one is handled, even no broadcast
+         */
+        if( !lwmac_timeout_is_running(&lwmac, TIMEOUT_WAIT_FOR_DEST_WAKEUP) ) {
+
+            /* Check if there are broadcasts to send and transmit immediately */
+            if(packet_queue_length(&(lwmac.tx.neighbours[0].queue)) > 0) {
+                lwmac.tx.current_neighbour = &(lwmac.tx.neighbours[0]);
+                lwmac_set_state(TRANSMITTING);
+                break;
+            }
+
+            lwmac_tx_neighbour_t* neighbour = _next_tx_neighbour(&lwmac);
+
+            if(neighbour != NULL) {
+                /* Offset in microseconds when the earliest (phase) destination
+                 * node wakes up that we have packets for. */
+                int time_until_tx = RTT_TICKS_TO_US(_ticks_until_phase(neighbour->phase));
+
+                /* If there's not enough time to prepare a WR to catch the phase
+                 * postpone to next interval */
+                if (time_until_tx < LWMAC_WR_PREPARATION_US) {
+                    time_until_tx += LWMAC_WAKEUP_INTERVAL_US;
+                }
+
+                time_until_tx -= LWMAC_WR_PREPARATION_US;
+                lwmac_set_timeout(&lwmac, TIMEOUT_WAIT_FOR_DEST_WAKEUP, time_until_tx);
+
+                /* Register neighbour to be the next */
+                lwmac.tx.current_neighbour = neighbour;
+
+                /* Stop dutycycling, we're preparing to send. This prevents the
+                 * timeout arriving late, so that the destination phase would
+                 * be missed. */
```

Comment:
@daniel-k Thanks a lot for the reply! This is clear to me now :-). But I still have two questions:

1) how quickly will the RIOT system (or the CPU) go into the sleep mode (totally stop) after setting up the lmp mode i.e., after executing ```lpm_prevent_sleep &= ~(LWMAC_LPM_MASK)```. Will RIOT sleep (stops running, all thread stops running) immediately after ```lpm_prevent_sleep &= ~(LWMAC_LPM_MASK)``` which in an extreme case that the next sentence after ```lpm_prevent_sleep &= ~(LWMAC_LPM_MASK)``` will not be executed??  Or, the OS will be in the sleep mode after a specific timing? I guess the OS will check the related sleeping flag in another thread and stop the CPU when it find the flag is set up, right?  

2) the second question is somehow related to the first question. Assume in a "no traffic" scene in LwMAC, LwMAC will only wake up (```lpm_prevent_sleep |= LWMAC_LPM_MASK```) in the wakeup period, and sleep (```lpm_prevent_sleep &= ~(LWMAC_LPM_MASK)```) in the sleep period. So the CPU and along with all the threads will only run in the wake up period (WP) and stops in the sleep period, right? like the following

|----WP---|----------SLEEP----------|----WP---|----------SLEEP----------|
||
V
| threads-|---------- threads--------|  threads-|---------- threads--------|
|---run----|-----------stops-----------|---run----|-----------stops-----------|      

I know that the lpm mode hasn't been completely implemented yet, I just want to know the way RIOT intend to work with lpm mode. Thanks!